ast/Helpers.h

```diff
@@ -546,16 +546,26 @@ class MK {
         return send->fun == core::Names::new_() && send->recv.isSelfReference();
     }
 
+    /*
+     * Is this an expression that refers to `::T`?
+     *
+     * This helper works on a purely syntactic level. For instance, this function determines if an expression is `T`,
+     * either with no scope or with the root scope (i.e. `::T`). This might not actually refer to the `T` that we define
+     * for users, but we don't know that information at the AST level.
+     */
     static bool isT(const ast::ExpressionPtr &expr) {
         bool result = false;
 
         typecase(
             expr,
             [&](const ast::UnresolvedConstantLit &t) {
+                // When the `T` was written by the user, we get an UnresolvedConstantLit.
                 result = t.cnst == core::Names::Constants::T() && ast::MK::isRootScope(t.scope);
             },
-            [&](const ast::ConstantLit &c) { result = c.symbol() == core::Symbols::T(); },
-            [&](const ast::ExpressionPtr &e) { result = false; });
+            [&](const ast::ConstantLit &c) {
+                // When the `T` was inserted by `ast::MK::T()`, we get a ConstantLit.
+                result = c.symbol() == core::Symbols::T();
+            });
```

Comment:
The comment here doesn't really apply anymore, because this function is no longer syntactic -- it also deals in resolved constants as well as syntactic constants.  It should be rewritten to reflect that fact.

I'm not sure that #8581 was necessarily the correct thing -- couldn't we have just made the RBS rewriter generate unresolved constants?  I get that that the resolved constant is slightly cheaper to pass through the rest of the pipeline, but the unresolved thing is more like what the user would be writing if they weren't writing RBS.

Reply:
We generate resolved constants because the `ast::MK::T` does so: https://github.com/sorbet/sorbet/blob/master/ast/Helpers.h#L369-L371.

Almost all `ast::MK` methods are using `T` as root. If we're going to change it it will impact the other rewriters in a lot of places.